src/newwave/main.mo

```diff
@@ -10,1382 +10,515 @@ import Iter "mo:base/Iter";
+    let fromEntityExists = checkIfEntityExists(bridgeToCreate.fromEntityId);
+
+    if (toEntityExists == false or fromEntityExists == false) {
+      return null;
     };
-  };
 
-  func createBridge(caller : Principal, bridgeToCreate : BridgeEntity.BridgeEntityInitiationObject) : async ?BridgeEntity.BridgeEntity {
-    // ensure that bridged Entities exist
-    switch(checkIfEntityWithAttributeExists("internalId", bridgeToCreate._fromEntityId)) {
-      case false { return null; }; // TODO: potentially return error message instead
-      case true {
-        if (checkIfEntityWithAttributeExists("internalId", bridgeToCreate._toEntityId) == false) {
-          return null; // TODO: potentially return error message instead
-        };
+    // Find a unique id for the new bridge that will not
+    // conflict with any current items
+    var newBridgeId : Text = "";
+    var counter : Nat = 0;
+    var found_unique_id : Bool = false;
+    while (not found_unique_id) {
+      // 10 is chosen arbitarily to ensure that in case of something weird happening
+      //  there is a timeout and it errors rather then looking forever
+      if (counter > 10) {
+        return null;
+      };
+
+      newBridgeId := await Utils.newRandomUniqueId();
+      if (bridgesStorage.get(newBridgeId) == null) {
+        let bridge : Bridge.Bridge = Bridge.generateBridgeFromInitializationObject(bridgeToCreate, newBridgeId, caller);
+        return addNewBridge(bridge);
       };
+
+      counter := counter + 1;
+    };
+    return null;
+  };
+
+  /**
+   * Function adds a new bridge and attempts to add the bridge to the entities
+   * bridge lookup tables. If adding the bridge to the entities fails, then the bridge
+   * is not created and is instead deleted
+   *
+   * @return Returns null if the entitiy failed to get created, otherwise it returns
+   * the bridge id of the newly created bridge
+  */
+  private func addNewBridge(bridge : Bridge.Bridge) : ?Text {
+    // Don't allow creating the bridge if the bridge already exists
+    if (checkIfBridgeExists(bridge.id) == true) {
+      return null;
     };
-    let bridge : BridgeEntity.BridgeEntity = await BridgeEntity.BridgeEntity(bridgeToCreate, caller);
-    let result = putBridge(bridge);
-    return ?result;
-  };
 
-  stable var bridgesStorageStable : [(Text, BridgeEntity.BridgeEntity)] = [];
-  var bridgesStorage : HashMap.HashMap<Text, BridgeEntity.BridgeEntity> = HashMap.HashMap(0, Text.equal, Text.hash);
+    // Add the bridge of the bridge database and add the bridge id to the related entities
+    let result = putBridge(bridge);
+    let fromIdResult = addBridgeToEntityFromIds(bridge.fromEntityId, bridge.id);
+    let toIdResult = addBridgeToEntityToIds(bridge.toEntityId, bridge.id);
+
+    // Ensure the bridge could be added to both entities bridge lookup tables
+    if (fromIdResult == false or toIdResult == false) {
+      // Delete the bridge since it failed to get created
```

Comment:
I think at this point we'll also delete any references that might have been created by either addBridgeToEntityFromIds or addBridgeToEntityToIds as one might have succeeded but the other failed

Reply:
Good catch